src/lib/mf.rs

```diff
@@ -509,16 +476,398 @@ fn rank_cnds<TokId: Clone + Copy + Debug + TryFrom<usize> + TryInto<usize> + Par
+                        Symbol::Term(t_idx) => {
+                            let off = usize::from(i) * terms_len + usize::from(t_idx);
+                            if table[off] != 0 {
+                                table[off] = 0;
+                                chgd = true;
+                            }
+                            term_cost(t_idx)
+                        }
+                    };
+
+                    for j in 0..terms_len {
+                        let this_off = usize::from(i) * terms_len + usize::from(j);
+                        let other_off = usize::from(sym_st_idx) * terms_len + usize::from(j);
+
+                        if table[other_off] != u64::max_value()
+                           && table[other_off] + d < table[this_off]
+                        {
+                            table[this_off] = table[other_off] + d;
+                            chgd = true;
+                        }
+                    }
+                }
+
+                // The second phase takes into account gotos.
+                for &(p_idx, sym_off) in sgraph.core_state(StIdx::from(i)).items.keys() {
+                    let prod = grm.prod(p_idx);
+                    if usize::from(sym_off) == prod.len() {
+                        for goto_idx in goto_edges.get(&(i, p_idx)).unwrap() {
+                            for j in 0..terms_len {
+                                let this_off = usize::from(i) * terms_len + usize::from(j);
+                                let other_off = usize::from(*goto_idx) * terms_len + usize::from(j);
+
+                                if table[other_off] != u64::max_value() &&
+                                    table[other_off] < table[this_off] {
+                                    table[this_off] = table[other_off];
+                                    chgd = true;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            if !chgd {
+                break;
+            }
+        }
+
+        Dist{terms_len, table}
+    }
+
+    pub(crate) fn dist(&self, st_idx: StIdx, t_idx: TIdx) -> Option<u64> {
+        let e = self.table[usize::from(st_idx) * self.terms_len + usize::from(t_idx)];
+        if e == u64::max_value() {
+            None
+        } else {
+            Some(e as u64)
+        }
+    }
+}
+
 #[cfg(test)]
 mod test {
     use std::convert::TryFrom;
+    use test::{Bencher, black_box};
 
-    use cfgrammar::yacc::YaccGrammar;
+    use cfgrammar::Symbol;
+    use cfgrammar::yacc::{yacc_grm, YaccGrammar, YaccKind};
     use lrlex::Lexeme;
+    use lrtable::{Minimiser, from_yacc, StIdx};
 
     use parser::{ParseRepair, RecoveryKind};
     use parser::test::do_parse;
-    use kimyi::test::{check_repairs, pp_repairs};
+    use kimyi::test::pp_repairs;
+
+    use super::Dist;
+
+    #[test]
+    fn dist_kimyi() {
+        let grms = "%start A
+%%
```

Comment:
You are using lower case terminals in your tests, so maybe this speaks for doing the same within the explanations.

Reply:
This test comes from the KimYi paper where they use lower-case terminal names. I agree it's a bit inconsistent, but I think it might be worse to diverge from the paper?